pkg/controller/dns/external_dns.go

```diff
@@ -0,0 +1,167 @@
+package dns
+
+import (
+	"time"
+
+	"github.com/Azure/aks-app-routing-operator/pkg/config"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/common"
+	"github.com/Azure/aks-app-routing-operator/pkg/manifests"
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	reconcileInterval = time.Minute * 3
+)
+
+// addExternalDnsReconciler creates a reconciler that manages external dns resources
+func addExternalDnsReconciler(manager ctrl.Manager, resources []client.Object) error {
+	return common.NewResourceReconciler(manager, "externalDnsReconciler", resources, reconcileInterval)
+}
+
+func addExternalDnsCleaner(manager ctrl.Manager, objs []cleanObj) error {
+	return nil // holding off on actually using this until e2e added in future pr
+
+	retriever := common.RetrieverEmpty()
+	for _, obj := range objs {
+		retriever = retriever.Add(common.RetrieverFromObjs(obj.resources, obj.labels)) // clean up entire unused external dns applications
+	}
+	for _, provider := range manifests.Providers {
+		labels := util.MergeMaps(provider.Labels(), manifests.TopLevelLabels)
+		retriever = retriever.Add(common.RetrieverFromGk(labels, manifests.OldExternalDnsGks...)) // clean up unused types from previous versions of app routing
+	}
+
+	retriever = retriever.Remove(common.RetrieverFromGk(
+		nil, // our compare strat is ignore labels
+		schema.GroupKind{
+			Group: corev1.GroupName,
+			Kind:  "Namespace",
+		}),
+		common.RemoveOpt{
+			CompareStrat: common.IgnoreLabels, // ignore labels, we never want to clean namespaces
+		})
+
+	return common.NewCleaner(manager, "externalDnsCleaner", retriever)
+}
+
+// NewExternalDns starts all resources required for external dns
+func NewExternalDns(manager ctrl.Manager, conf *config.Config, self *appsv1.Deployment) error {
+	instances := instances(conf, self)
+
+	deployInstances := filterAction(instances, deploy)
+	deployRes := getResources(deployInstances)
+	if err := addExternalDnsReconciler(manager, deployRes); err != nil {
+		return err
+	}
+
+	cleanObjs := cleanObjs(instances)
+	if err := addExternalDnsCleaner(manager, cleanObjs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func instances(conf *config.Config, self *appsv1.Deployment) []instance {
+	// public
+	publicConfig := publicConfig(conf)
+	publicAction := deploy
+	if len(conf.PublicZoneConfig.ZoneIds) == 0 {
+		publicAction = clean
+	}
+	publicResources := manifests.ExternalDnsResources(conf, self, []*manifests.ExternalDnsConfig{publicConfig})
+
+	// private
+	privateConfig := privateConfig(conf)
+	privateAction := deploy
```

Comment:
I moved this into a separate function. Good idea!

We can't move it into the config functions because we want to adhere to the single responsibility principle. The config functions should only return a config.